homeassistant/components/sensor/integration.py

```diff
@@ -0,0 +1,180 @@
+"""
+Numeric integration of data coming from a source sensor over time.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.integration/
+"""
+import logging
+
+from decimal import Decimal, DecimalException
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_NAME, ATTR_UNIT_OF_MEASUREMENT, STATE_UNKNOWN, STATE_UNAVAILABLE)
+from homeassistant.core import callback
+from homeassistant.helpers.event import async_track_state_change
+from homeassistant.helpers.restore_state import RestoreEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_SOURCE_ID = 'source'
+
+CONF_SOURCE_SENSOR = 'source'
+CONF_ROUND_DIGITS = 'round'
+CONF_UNIT_PREFIX = 'unit_prefix'
+CONF_UNIT_TIME = 'unit_time'
+CONF_UNIT_OF_MEASUREMENT = 'unit'
+
+# SI Metric prefixes
+UNIT_PREFIXES = {None: 1,
+                 "k": 10**3,
+                 "G": 10**6,
+                 "T": 10**9}
+
+# SI Time prefixes
+UNIT_TIME = {'s': 1,
+             'min': 60,
+             'h': 60*60,
+             'd': 24*60*60}
+
+ICON = 'mdi:char-histogram'
+
+DEFAULT_ROUND = 3
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_SOURCE_SENSOR): cv.entity_id,
+    vol.Optional(CONF_ROUND_DIGITS, default=DEFAULT_ROUND): vol.Coerce(int),
+    vol.Optional(CONF_UNIT_PREFIX, default=None): vol.In(UNIT_PREFIXES),
+    vol.Optional(CONF_UNIT_TIME, default='h'): vol.In(UNIT_TIME),
+    vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string
+})
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the integration sensor."""
+    integral = IntegrationSensor(config[CONF_SOURCE_SENSOR],
+                                 config.get(CONF_NAME),
+                                 config[CONF_ROUND_DIGITS],
+                                 config[CONF_UNIT_PREFIX],
+                                 config[CONF_UNIT_TIME],
+                                 config.get(CONF_UNIT_OF_MEASUREMENT))
+
+    async_add_entities([integral])
+
+
+class IntegrationSensor(RestoreEntity):
+    """Representation of an integration sensor."""
+
+    def __init__(self, source_entity, name, round_digits, unit_prefix,
+                 unit_time, unit_of_measurement):
+        """Initialize the integration sensor."""
+        self._sensor_source_id = source_entity
+        self._round_digits = round_digits
+        self._state = 0
+
+        self._name = name if name is not None\
+            else '{} integral'.format(source_entity)
+
+        if unit_of_measurement is None:
+            self._unit_of_measurement = "" if unit_prefix is None\
+                                        else unit_prefix
+            self._unit_of_measurement += "{}" + unit_time
```

Comment:
Use new style string formatting instead of string concatenation for the last local variable.